### PR TITLE
refactor(discovery): unify duplicated discovery loops and harden DeviceService parsing

### PR DESCRIPTION
**Pull request by Djelibeybi**
## Summary

This branch bundles the post-v1.0 `/simplify` work accumulated since `main` (2026-06-11), across three logical areas. All gates green: **2511 tests pass**, `ruff` clean, `pyright` 0/0/0.

> Note on scope: this is a larger-than-usual branch covering three related-but-distinct areas (below). Happy to split into separate PRs if preferred for review.

### 1. Unify duplicated discovery loops (the main change)

Rebuilds `discover_devices()` on top of the shared `_discover_with_packet()` async generator, removing ~150 near-identical lines that had already drifted apart:

- **Hoisted DoS protections** into the shared generator so *every* discovery caller (`discover`, `find_by_serial`, `find_by_label`, `find_by_ip`) inherits them — serial/broadcast-bit/all-zeros validation, source-ID checks, and first-wins per-serial dedup.
- **Extracted `IdleDeadline`** (`network/utils.py`) — the dual overall+idle timeout arithmetic that was hand-rolled identically in three loops, now one tested helper.
- **mDNS** `discover_lifx_services` adopts `IdleDeadline` and routes `receive()` exceptions by type (genuine socket errors no longer masquerade as "no devices").
- Retired the hand-rolled `_parse_device_state_service()` in favour of `StateService.unpack()`.
- Deprecated `UdpTransport.receive_many` (→ `v6.0`, points at the public `lifx.api`).

### 2. Real-hardware discovery robustness (found during validation)

Discovery against real devices surfaced `ValueError: 5 is not a valid DeviceService` — devices advertise one `StateService` per service they support, including identifiers newer than the bundled `protocol.yml` defines. The emulator only ever sends `service=UDP(1)`, so tests never caught it. Fixed at two layers:

- **`protocol/base.py`** — enum deserialisation now tolerates unknown values (falls back to the raw int via `_coerce_enum`) instead of raising. General fix: applies to *any* enum field from newer firmware, and removes a log-amplification path (the prior `ValueError` was logged at ERROR with a full traceback per packet).
- **`network/discovery.py`** — the `GetService` path ignores non-UDP `StateService` replies so they neither claim the serial via dedup nor supply a non-UDP port. Matches every mature LIFX client.

### 3. Ceiling save-on-exit + transport simplification

- `CeilingLight.__aexit__` persists current state before `close()` (final save then close, including on exception paths), via `asyncio.to_thread` to keep file I/O off the loop.
- `/simplify` dedup of UDP transport mechanics and connection retry-budget hot paths.

## Verification

- `uv run --frozen pytest` — **2511 passed, 12 deselected**
- `uv run ruff format . && uv run ruff check .` — clean
- `uv run pyright` — 0 errors, 0 warnings
- Real-hardware discovery confirmed working by the maintainer (`examples/discovery_broadcast.py`).
- Phase artifacts under `.planning/phases/01-unify-duplicated-discovery-loops/`: code review (6/6 fixed), security (`01-SECURITY.md`, 11/11 threats closed), UAT (4/4 incl. hardware).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/lifx/network/discovery.py [hunk 1]
```diff
@@ -282,6 +299,22 @@ async def _discover_with_packet(
                     )
                     continue
 
+                # Reject broadcast/multicast serials (D-01, D-02). The
+                # multicast bit check also covers the all-0xff broadcast
+                # target; the all-zeros target is the LIFX broadcast address
+                # used by the discovery request itself and is never a valid
+                # device serial.
+                if header.target[0] & 0x01 or header.target == b"\x00" * 8:
```
Comment (Djelibeybi) [inline src/lifx/network/discovery.py]: Warning: This now rejects multicast/broadcast and all-zero targets, but it still accepts non-zero bytes in the two protocol padding positions. `Serial.from_protocol()` drops `header.target[6:]`, so a malformed response like `d073d5010203ffff` is normalized to `d073d5010203` and can win the per-serial dedup before the real device replies. Since the request source is visible on the LAN, a spoofed response can poison discovery with the attacker's IP/port. Please also reject non-zero padding, e.g. `or header.target[6:] != b"\x00\x00"`, and add a regression test for valid first six bytes plus non-zero padding.

File: src/lifx/devices/ceiling.py [hunk 2]
```diff
@@ -204,6 +207,31 @@ async def __aenter__(self) -> CeilingLight:
...
+        delegating to the parent ``close()`` via ``super().__aexit__()``.  The
+        save runs in a worker thread via ``asyncio.to_thread`` so the file I/O
+        never blocks the event loop.  It is guarded by a belt-and-braces
+        ``try/except`` so that any I/O failure is logged as a WARNING and
+        swallowed — the original body exception (if any) is never replaced or
+        suppressed, and the connection is always cleaned up.
+        """
+        if self._state_file:
+            try:
+                await asyncio.to_thread(self._save_state_to_file)
```
Comment (Djelibeybi) [inline src/lifx/devices/ceiling.py]: Warning: Cleanup is not protected if cancellation lands while the threaded save is pending. `asyncio.CancelledError` is a `BaseException`, so it bypasses the `except Exception` block and line 233 never runs, leaving the connection open despite the docstring's 'always cleaned up' guarantee. Wrap the save attempt in an outer `try/finally` and call `await super().__aexit__(...)` from the `finally`, while keeping the inner `except Exception` for ordinary save failures.

**Review by Djelibeybi — COMMENTED** [review]
## Hermes Agent Review

Found 2 correctness issues worth fixing before merge; posted inline. Because this review is being submitted by the PR author account, I'm using a COMMENT review rather than REQUEST_CHANGES, but these are changes-requested-in-spirit.

Verification performed locally:
- `python -m pytest tests/test_network/test_discovery_errors.py tests/test_protocol/test_base.py tests/test_devices/test_state_ceiling.py tests/test_network/test_connection.py tests/test_network/test_mdns/test_discovery.py tests/test_network/test_mdns/test_transport.py tests/test_network/test_transport.py tests/test_network/test_utils.py` -> 166 passed, 7 expected deprecation warnings.
- `python -m ruff check ...` on touched implementation/focused test files -> All checks passed.

CI is also green on the PR head.

File: src/lifx/network/discovery.py (hunk 1 above)
Comment (Djelibeybi) [reply src/lifx/network/discovery.py]: Good catch — fixed in 3b97ee3. The guard now also rejects non-zero padding (`or header.target[6:] != b"\x00\x00"`), since `Serial.from_protocol()` drops `target[6:]` and a non-zero-padded spoof would otherwise normalise to a clean serial and win the dedup race. Added a regression test (`TestNonZeroTargetPadding`: valid first six bytes + `ffff` padding → rejected, no device yielded). Real devices always send zero padding, so no legitimate response is affected.

File: src/lifx/devices/ceiling.py (hunk 2 above)
Comment (Djelibeybi) [reply src/lifx/devices/ceiling.py]: Confirmed and fixed in 1875dbc. The save is now wrapped in `try/finally` with `await super().__aexit__(...)` in the `finally`, so the parent cleanup runs even when `asyncio.CancelledError` (a `BaseException`) lands during the threaded save; the inner `except Exception` still swallows ordinary save failures, and the cancellation still propagates afterwards. Added a regression test that patches `asyncio.to_thread` to raise `CancelledError` and asserts `close()` still runs while the `CancelledError` propagates.

**Review by Djelibeybi — COMMENTED** [review]
## Hermes Agent Re-review

The two previous inline findings look addressed:

- `src/lifx/network/discovery.py` now rejects non-zero target padding before `Serial.from_protocol()`, with a regression test covering the spoofed-padding case.
- `src/lifx/devices/ceiling.py` now wraps the save path in `try/finally` so parent cleanup runs even if cancellation lands during the threaded save, with a regression test covering that path.

No new line-specific findings from this pass, so I am not posting fresh inline comments.

Local verification:
- Focused suite: `python -m pytest tests/test_network/test_discovery_errors.py tests/test_protocol/test_base.py tests/test_devices/test_state_ceiling.py tests/test_network/test_connection.py tests/test_network/test_mdns/test_discovery.py tests/test_network/test_mdns/test_transport.py tests/test_network/test_transport.py tests/test_network/test_utils.py` -> 168 passed, 7 expected deprecation warnings.
- Lint: `python -m ruff check .` -> All checks passed.
- Full local suite on macOS / Python 3.14.5: `python -m pytest` -> 2524 passed, 12 deselected, 7 expected deprecation warnings.

One caveat: GitHub CI currently has one failing job, `Test (Python 3.14 on ubuntu-latest)`, with `tests/test_api/test_api_apply_theme.py::TestDeviceGroupApplyTheme::test_apply_theme_basic` timing out against `127.0.0.1`. The same full suite passed locally on Python 3.14.5, and the other matrix jobs are green, so this smells like an emulator/network flake rather than a fresh code-review finding — but I would rerun or clear that CI failure before merging.
